### PR TITLE
head change coalescer

### PR DESCRIPTION
**Pull request by vyzo**

### CODE REVIEW

File: chain/store/coalescer.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	for _, ts := range revert {
+		key := ts.Key()
+
+		_, ok := applied[key]
+		if ok {
+			reverted[key] = struct{}{}
+			continue
+		}
+
+		c.revert = append(c.revert, ts)
```
Comment (whyrusleeping) [inline chain/store/coalescer.go:104]: it looks like we never remove anything from the revert queue, which seems to imply to me that we could end up with the same thing in the revert queue multiple times

Comment (vyzo) [reply chain/store/coalescer.go:104]: why would it be reverted multiple times however? There has to be an apply in between, and those will cancel, right?

Comment (vyzo) [reply chain/store/coalescer.go]: Implemented coalescing logic for both revert and apply, as we discussed synchronously.

File: chain/store/coalescer.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	// - pending reverts are cancelled by incoming applys
+	// - incoming reverts are cancelled by pending applys
+	newRevert := make([]*types.TipSet, 0, len(c.revert)+len(revert))
+	for _, ts := range c.revert {
+		_, cancel := applying[ts.Key()]
+		if cancel {
+			continue
+		}
+
+		newRevert = append(newRevert, ts)
```
Comment (arajasek) [inline chain/store/coalescer.go:133]: Nit: Can these be changed to 

```
if !cancel {
    newRevert = append(newRevert, ts)
}
```

**Review by arajasek — APPROVED** [review]
Logic looks good to me!

**Comment by whyrusleeping** [comment]
@vyzo so the code here appears to work pretty well, but I'm still seeing the mempool triggering recent state calls. Maybe we can just bump the timeout number? Will try now...

**Comment by whyrusleeping** [comment]
2020-11-06-135712_779x1810_scrot

I bumped the delay up to 3 seconds, and ran this for a bit. Its *better*, but i'm still seeing state computations being triggered by the mempool (and if this were fully fixed, i'd expect to see them only be triggered by the FundManager for now)

**Comment by whyrusleeping** [comment]
To check this, I just grabbed a random 120 second profile of the lotus node while its fully caught up

**Comment by vyzo** [comment]
We can try to add some more smarts -- for instance, when it's time to trigger the callback check to see if the last coalesce was within some bound (say 100ms) -- and if that's the case wait some more.
This might succeed in coalescing all the incoming blocks in the beginning of the epoch.

**Comment by vyzo** [comment]
Also, _some_ state computations are unavoidable, we do have to compute at some point, we would just like to avoid the unnecessary ones :)

**Comment by vyzo** [comment]
Added some more smarts to the coalescing algorithm, controlled by 3 parameters: minDelay, maxDelay, and mergeInterval.

When the first head chance (since the last dispatch) is received, the coalescer schedules a timer for minDelay to coalesce more head changes.
When the timer fires, it checks the time of the last coalesce; if it is within the mergeInterval, then it extends the coalesce time to coalesce some more.
Finally, the delivery of head change is not delaye more than maxDelay.

For the mpool, I have set minDelay=2s, maxDelay=6s, and mergeInterval=1s.
Hopefully this will capture the common case of 6s or less for block delivery.

**Review by arajasek — APPROVED** [review]
New version looks good to me

**Review by Kubuxu — APPROVED** [review]

**Review by magik6k — APPROVED** [review]

**Pull request merged**